# Курсовая работа. Блок 1
*Pandas и Matplotlib*

**NB!** Код должен быть с комментариями к каждому логическому блоку кода. В противном случае работа **не** будет принята. <br><br>
Выполнение курсовой работы может потребовать использование дополнительных библиотек. 

<br><br>
**Описание работы:**

Перед вами стоит бизнес-задача – на основании имеющихся данных подготовить аналитический отчет, который в дальнейшем поможет продюсерам образовательных программ эффективно выстраивать стратегию по модернизированию и улучшению курсов. В начале отчета предлагается оформить емкий описательный блок по каждому курсу на основании рассчитанных показателей. Далее предлагается посчитать потенциальную нагрузку на преподавателей, чтобы оценить необходимость расширения штата сотрудников. Затем идет блок из двух пунктов по анализу качества контента курсов, где необходимо выявить проблемные модули, которые, возможно, требуют доработки. Также стоит задача выявить потенциальную сезонность. Наконец, предложено задание для самостоятельной разработки метрики успеваемости студентов для нахождения тех, кто значительно хуже справляются с прохождением курса. Каждый из пунктов анализа предполагается сопроводить аналитическим выводом на основании рассчитанных метрик.

<br><br>

_________
Обозначения:<br><br>
&nbsp;&nbsp;&nbsp;&nbsp;**(p)** – задание может быть выполнено после прохождения модулей по Pandas <br>
&nbsp;&nbsp;&nbsp;&nbsp;**(m)** – задание может быть выполнено после прохождения модуля по Matplotlib <br>
&nbsp;&nbsp;&nbsp;&nbsp;⭐ – необязательное задание повышенной сложности
_________

[Codebook](#Codebook) <br>
[1. Описание и начальная работа с данными](#1.1-Описание-и-начальная-работа-с-данными)<br>
[2. Расчет потенциальной нагрузки на преподавателей](#2.-Расчет-потенциальной-нагрузки-на-преподавателей)<br>
[3. Выявление проблемных модулей](#3.-Выявление-проблемных-модулей)<br>
[4. Расчет конверсии](#4.-Расчет-конверсии) <br>
[5. Метрика успеваемости ](#5.-Метрика-успеваемости)

## Codebook

`courses.csv` содержит следующие значения: <br><br>
&nbsp;&nbsp;&nbsp;&nbsp; `id` – идентификатор курса <br>
&nbsp;&nbsp;&nbsp;&nbsp; `title` – название курса <br>
&nbsp;&nbsp;&nbsp;&nbsp;  `field` – сфера, к которой относится курс <br> <br><br>
`students.csv` содержит следующие значения: <br><br>
&nbsp;&nbsp;&nbsp;&nbsp; `id` – идентификатор студента <br>
&nbsp;&nbsp;&nbsp;&nbsp; `city` – город студента <br>
&nbsp;&nbsp;&nbsp;&nbsp;  `birthday` – день рождения студента <br> <br><br>
`course_contents.csv` содержит следующие значения: <br><br>
&nbsp;&nbsp;&nbsp;&nbsp; `course_id` – идентификатор курса <br>
&nbsp;&nbsp;&nbsp;&nbsp; `module_number` – номер модуля <br>
&nbsp;&nbsp;&nbsp;&nbsp;  `module_title` – название модуля <br> 
&nbsp;&nbsp;&nbsp;&nbsp; `lesson_number` – номер урока <br>
&nbsp;&nbsp;&nbsp;&nbsp;  `lesson_title` – название урока <br>
&nbsp;&nbsp;&nbsp;&nbsp;  `lesson_token` – токен урока <br> 
&nbsp;&nbsp;&nbsp;&nbsp; `is_video` – наличие видео *(true/false)* <br>
&nbsp;&nbsp;&nbsp;&nbsp;  `is_homework` – наличие домашней работы *(true/false)* <br>
<br><br>
`progresses.csv` содержит следующие значения: <br><br>
&nbsp;&nbsp;&nbsp;&nbsp; `id` – идентификатор прогресса <br>
&nbsp;&nbsp;&nbsp;&nbsp; `student_id` – идентификатор студента <br>
&nbsp;&nbsp;&nbsp;&nbsp;  `course_id` – идентификатор курса <br> <br><br>
`progress_phases.csv` содержит следующие значения: <br><br>
&nbsp;&nbsp;&nbsp;&nbsp; `progress_id` – идентификатор прогресса <br>
&nbsp;&nbsp;&nbsp;&nbsp; `module_number` – номер модуля <br>
&nbsp;&nbsp;&nbsp;&nbsp; `lesson_number` – номер урока <br>
&nbsp;&nbsp;&nbsp;&nbsp;  `status` – статус прохождения урока <br>
&nbsp;&nbsp;&nbsp;&nbsp;  `start_date` – дата начала <br> 
&nbsp;&nbsp;&nbsp;&nbsp; `finish_date` – дата окончания <br>
<br><br>

## 1. Описание и начальная работа с данными

Вам необходимо подготовить данные и описать их. Данные реальные и содержат пропущенные значения, а также лишние относительно друг друга данные. <br>

Объедините датасеты (кроме `students.scv`) в один общий для дальнейшей работы. Отдельный датасет создайте для расчета среднего возраста студентов курсов (отдельный датасет со `students.scv` нужен, чтобы в основном не потерять значения в основном датасете из-за того, что не по всем студентам есть анкетные данные о дате рождения). <br> <br>
Опишите данные: <br>
1. **(p)** Посчитайте
      * общее количество курсов в датасете, 
      * количество модулей на каждом курсе, 
      * количество уроков в каждом модуле на каждом курсе, 
      * медианное количество уроков в модуле на каждом курсе, 
      * количество учеников на каждом курсе
      * минимальный, максимальный, средний, медианный возраст студентов
      * минимальный, максимальный, средний, медианный возраст студентов на каждом курсе
2. **(m)** Постройте bar-chart, отражающий количество студентов на каждом курсе. Ticks нужно развернуть так, чтобы они были читаемы
3. **(m)** Постройте горизонтальный (столбцы должны располагаться горизонтально) bar-chart, отражающий количество студентов на каждом курсе. График должен иметь заголовок. Значения должны быть отсортированы. Цвет столбцов должен содержать информацию о сфере, к которой относится курс (то есть нужна легенда). Прозрачность должна стоять на отметке 0.1. На график должна быть нанесена линия медианы. У медианы должен быть свой цвет. Рамки у графика быть не должно ⭐
4.     На основании рассчитанных значений опишите данные (описание должно быть полным и покрывать все полученные выше метрики)

_____________________________________________________________________


Для того, чтобы проверить, что вы правильно соединили все датасеты в основной (без `students.csv`), запустите аналогичный приведенному ниже запрос. Выдача должна совпадать. Имейте в виду, что названия некоторых столбцов были переименованы.

In [2]:
# main[main.student_id == '768c2987a744c51ce64a5993a2a94eaf'] # проверка основного датасета

progress_id  module_number  lesson_number status  \
0       a387ab916f402cb3fbfffd29f68fd0ce              2              4   done   
1       a387ab916f402cb3fbfffd29f68fd0ce              1              1   done   
2       a387ab916f402cb3fbfffd29f68fd0ce              2              2   done   
3       a387ab916f402cb3fbfffd29f68fd0ce              1              4   done   
4       a387ab916f402cb3fbfffd29f68fd0ce              1              6   done   
5       a387ab916f402cb3fbfffd29f68fd0ce              2              1   done   
6       a387ab916f402cb3fbfffd29f68fd0ce              1              3   done   
7       a387ab916f402cb3fbfffd29f68fd0ce              2              5  start   
8       a387ab916f402cb3fbfffd29f68fd0ce              1              2   done   
9       a387ab916f402cb3fbfffd29f68fd0ce              1              5   done   
10      a387ab916f402cb3fbfffd29f68fd0ce              2              3   done   
112420  6407c290505e790e92207a1dbc41a2cd              1              1  start   
132870  5767046d89ed7563a0c3a9a3896a23e6              1              3   done   
132871  5767046d89ed7563a0c3a9a3896a23e6              2              2   done   
132872  5767046d89ed7563a0c3a9a3896a23e6              1              1   done   
132873  5767046d89ed7563a0c3a9a3896a23e6              1              2   done   
132874  5767046d89ed7563a0c3a9a3896a23e6              2              1   done   
132875  5767046d89ed7563a0c3a9a3896a23e6              2              3   fail   
155020  c90ebe1431eac5cbb11692100b7a0f8d              1              3   done   
155021  c90ebe1431eac5cbb11692100b7a0f8d              1              1   done   
155022  c90ebe1431eac5cbb11692100b7a0f8d              1              7  start   
155023  c90ebe1431eac5cbb11692100b7a0f8d              1              5   done   
155024  c90ebe1431eac5cbb11692100b7a0f8d              1              6   done   
155025  c90ebe1431eac5cbb11692100b7a0f8d              1              4   done   
155026  c90ebe1431eac5cbb11692100b7a0f8d              1              2   done   

                           start_date                    finish_date  \
0       2018-06-23 08:28:50.681065+00  2018-06-23 08:28:52.439542+00   
1       2018-06-20 14:25:21.783762+00  2018-06-20 15:45:07.717209+00   
2       2018-06-23 08:18:09.653771+00  2018-06-23 08:18:12.784616+00   
3        2018-06-20 16:00:06.36178+00  2018-06-21 19:09:30.845034+00   
4       2018-06-21 19:10:36.957891+00  2018-06-28 15:59:25.320418+00   
5       2018-06-23 08:18:07.476199+00  2018-06-23 08:18:09.552181+00   
6       2018-06-20 15:58:22.132468+00  2018-06-20 16:00:06.127191+00   
7       2018-06-23 08:31:00.338883+00                            NaN   
8       2018-06-20 15:45:07.976223+00  2018-06-20 15:58:21.886722+00   
9       2018-06-21 19:09:31.117521+00  2018-06-21 19:10:36.707919+00   
10      2018-06-23 08:18:12.890698+00  2018-06-23 08:28:50.564437+00   
112420  2018-06-20 14:25:13.010259+00                            NaN   
132870  2018-02-19 12:32:35.928905+00  2018-02-20 09:16:36.741375+00   
132871  2018-02-22 08:57:31.008346+00  2018-04-28 13:45:40.708228+00   
132872         2018-01-13 15:20:23+00         2018-01-13 15:20:23+00   
132873         2018-01-13 15:20:23+00         2018-01-14 13:47:25+00   
132874  2018-02-22 08:54:10.987784+00  2018-02-22 08:57:30.914939+00   
132875  2018-04-28 13:45:40.955076+00                            NaN   
155020  2018-06-19 15:08:14.103923+00  2018-06-19 15:39:53.661163+00   
155021  2018-06-19 10:40:05.063485+00  2018-06-19 14:56:16.346353+00   
155022  2018-06-19 18:40:00.576904+00                            NaN   
155023  2018-06-19 18:10:52.316728+00  2018-06-19 18:35:38.720317+00   
155024  2018-06-19 18:35:38.884217+00  2018-06-19 18:37:11.155066+00   
155025  2018-06-19 15:39:53.923777+00    2018-06-19 18:10:52.1737+00   
155026  2018-06-19 14:56:16.570129+00  2018-06-19 15:08:13.930725+00   

                              student_id         

In [3]:
import pandas as pd
# import dask.dataframe as dd

# импорт всех необходимых csv-файлов
students = pd.read_csv('students.csv', parse_dates=['birthday'])
courses = pd.read_csv('courses.csv').drop('Unnamed: 0', axis=1)
courses.columns = ['id', 'course_title', 'course_field']
course_contents = pd.read_csv('course_contents.csv')
progresses = pd.read_csv('progresses.csv')
progress_phases = pd.read_csv('progress_phases.csv')

In [4]:
# создание одного общего датасета
df = pd.merge(progress_phases, progresses, right_on='id', left_on='progress_id', how='left').drop('id', axis=1)
df = pd.merge(df, courses, left_on='course_id', right_on='id', how='left').drop('id', axis=1)
df = pd.merge(df,
              course_contents[['course_id', 'module_title', 'lesson_title', 'lesson_token', 'is_video', 'is_homework',
                               'module_number', 'lesson_number']],
              left_on=['course_id', 'module_number', 'lesson_number'],
              right_on=['course_id', 'module_number', 'lesson_number'],
              how='inner',
              suffixes=('', '_y'))
df.drop(df.filter(regex='_y$').columns.tolist(), axis=1, inplace=True)

In [4]:
# датасет для расчета среднего возраста студентов
students_birthday = students.dropna(subset=['birthday']).reset_index(drop=True)

**1.1 Общее количество курсов:**

In [5]:
len(courses.index)

15

**1.2 Количество модулей в каждом курсе:**

In [131]:
df.groupby('course_title')['module_title'].nunique().reset_index().sort_values(by='module_title', ascending=False)

course_title  module_title
7                  Анимация интерфейсов            21
6                             UX-дизайн            20
12                      Веб-разработчик            20
11                Веб-дизайн с нуля 2.0            19
3                     JavaScript с нуля            18
13        Интернет-маркетолог от Ingate            18
1                      Java-разработчик            17
9                    Веб-дизайн PRO 2.0            17
10                   Веб-дизайн Базовый            17
14        Руководитель digital-проектов            17
5              SMM-маркетолог от А до Я            11
0                         Excel Базовый             9
2               Java-разработчик c нуля             9
4   PHP-разработчик с 0 до PRO. Часть 1             8
8        Веб-вёрстка для начинающих 2.0             8

**1.3 Количество уроков в каждом модуле на каждом курсе:**

In [133]:
df_lesson_count = df.groupby(['course_title', 'module_title'])['lesson_title'].nunique().reset_index()
df_lesson_count.sort_values(by='lesson_title', ascending=False)

course_title  \
194        Интернет-маркетолог от Ingate   
34               Java-разработчик c нуля   
195        Интернет-маркетолог от Ingate   
200        Интернет-маркетолог от Ingate   
207        Интернет-маркетолог от Ingate   
..                                   ...   
192                      Веб-разработчик   
193                      Веб-разработчик   
64              SMM-маркетолог от А до Я   
56   PHP-разработчик с 0 до PRO. Часть 1   
111                 Анимация интерфейсов   

                                          module_title  lesson_title  
194                                    Email-маркетинг            29  
34                                Числа, строки и даты            27  
195  SEO: больше, чем поисковая выдача (обновление ...            19  
200  Контекстная реклама в Yandex Direct и Google A...            17  
207         Таргетированная реклама в социальных сетях            16  
..                                                 ...           ...  
192                                           Проверка             1  
193                            Экзаменационная работа              1  
64                      Информация о дипломном проекте             1  
56                                     Курсовая работа             1  
111                      Рабочие будни моушн-дизайнера             1  

[229 rows x 3 columns]

**1.4 Медианное количество уроков в модуле на каждом курсе:**

In [126]:
df_lesson_count.groupby(['course_title'])['lesson_title'].aggregate(['median']).sort_values(by='median', 
                                                                                            ascending=False).reset_index()

course_title  median
0               Java-разработчик c нуля    10.0
1                      Java-разработчик     7.0
2                     JavaScript с нуля     7.0
3        Веб-вёрстка для начинающих 2.0     7.0
4         Интернет-маркетолог от Ingate     6.5
5              SMM-маркетолог от А до Я     6.0
6                         Excel Базовый     5.0
7                    Веб-дизайн PRO 2.0     5.0
8         Руководитель digital-проектов     5.0
9   PHP-разработчик с 0 до PRO. Часть 1     4.0
10                Веб-дизайн с нуля 2.0     4.0
11                            UX-дизайн     3.5
12                 Анимация интерфейсов     3.0
13                   Веб-дизайн Базовый     3.0
14                      Веб-разработчик     2.0

**1.5 Количество учеников на каждом курсе**

In [134]:
count_students = df.groupby('course_title')['student_id'].nunique().reset_index().sort_values(by='student_id', ascending=False)
count_students

course_title  student_id
13        Интернет-маркетолог от Ingate        2168
11                Веб-дизайн с нуля 2.0        2014
8        Веб-вёрстка для начинающих 2.0        2004
9                    Веб-дизайн PRO 2.0        1711
6                             UX-дизайн        1151
3                     JavaScript с нуля         966
4   PHP-разработчик с 0 до PRO. Часть 1         854
0                         Excel Базовый         782
1                      Java-разработчик         763
14        Руководитель digital-проектов         685
12                      Веб-разработчик         628
7                  Анимация интерфейсов         598
2               Java-разработчик c нуля         581
10                   Веб-дизайн Базовый         518
5              SMM-маркетолог от А до Я         511

**1.6 Минимальный, максимальный, средний, медианный возраст студентов:**

In [23]:
import datetime

# принудительно приводим столбец "birthday" к формату datetime (пропущенные значения=NaT)
students_birthday['birthday'] = pd.to_datetime(students_birthday['birthday'], errors='coerce')

# устанавливаем ограничение по возрасту, так как есть аномальные значения (человек не может быть возраста: 1 год, 100 и 
# более лет). Пусть будет от 12 до 100 лет.
students_birthday_red = students_birthday[(students_birthday['birthday'] > '1920') & (students_birthday['birthday'] <= '2008')]
# вычисляем возраст, как разницу между сегодняшней датой и датой рождения, результат получается в днях, делим на 365, 
# получаем год. Результат сохраняем в новый стлобец "age".
students_birthday_red['age'] = (datetime.datetime.now() - students_birthday_red['birthday']).apply(lambda x: (x / 365).days)
students_birthday_red['age'].agg(['min', 'max', 'mean', 'median']).reset_index()


C:\AnacondaNew\lib\site-packages\ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if __name__ == '__main__':


index         age
0     min   12.000000
1     max  100.000000
2    mean   29.643037
3  median   29.000000

**1.7 Минимальный, максимальный, средний, медианный возраст студентов на каждом курсе**

In [51]:
df_age = pd.merge(df[['student_id', 'course_title']], students_birthday_red[['id', 'age']], 
                  left_on='student_id', right_on='id', how='inner')
# удаляем дубликаты строк
df_age = df_age.drop_duplicates(subset=['student_id', 'course_title'])

In [50]:
df_age.groupby('course_title')['age'].agg(['min', 'max', 'mean', 'median'])

min  max       mean  median
course_title                                                    
Excel Базовый                         17   61  34.180180      34
Java-разработчик                      15   65  29.876274      29
Java-разработчик c нуля               14   55  30.070690      29
JavaScript с нуля                     14   59  29.142232      28
PHP-разработчик с 0 до PRO. Часть 1   14   75  30.053416      29
SMM-маркетолог от А до Я              17   52  29.573441      29
UX-дизайн                             15   85  30.744233      30
Анимация интерфейсов                  15   51  30.162011      30
Веб-вёрстка для начинающих 2.0        14   59  29.162178      28
Веб-дизайн PRO 2.0                    15   59  29.009410      28
Веб-дизайн Базовый                    16   51  29.119681      28
Веб-дизайн с нуля 2.0                 14   76  28.734555      28
Веб-разработчик                       14   51  28.678038      28
Интернет-маркетолог от Ingate         17   90  31.786022      31
Руководитель digital-проектов         19   55  32.109123      32

**2. Постройте bar-chart, отражающий количество студентов на каждом курсе. Ticks нужно развернуть так, чтобы они были читаемы**

In [60]:
import plotly.express as px
fig1 = px.bar(count_students, x=count_students['course_title'], y=count_students['student_id'])
fig1.update_layout(legend_orientation='h', legend=dict(x=.5, xanchor='center'), 
                   title='<b>Количество студентов на каждом курсе</b>')
fig1.update_xaxes(title='Курсы')
fig1.update_yaxes(title='Количество студентов')
fig1.show()

**3. Постройте горизонтальный (столбцы должны располагаться горизонтально) bar-chart, отражающий количество студентов на каждом курсе.**

In [71]:
# добавляем к табличке count_students столбец "сферы"
count_students = count_students.merge(courses[['course_field', 'course_title']], left_on='course_title', 
                                      right_on='course_title', suffixes = ['course_title', 'title'])

In [101]:
median_count = count_students['student_id'].median()

In [119]:
import plotly.graph_objs as go
fig2 = px.bar(count_students, x=count_students['student_id'], y=count_students['course_title'],
              color=count_students['course_field'])
fig2.update_layout(legend_orientation='h', legend=dict(x=.5, xanchor='center'),
                   title={'text': '<b>Количество студентов на курсах</b>',
                          'x': 0.5,
                          'xanchor': 'center'},
                   xaxis_title='Количество студентов',
                   yaxis_title='Курсы',
                   template='simple_white')
fig2.update_yaxes(categoryorder='total ascending')
fig2.update_traces(opacity=0.4)

fig2.add_traces(go.Scatter(x=[median_count]*count_students.shape[0], y=count_students['course_title'], mode='lines',
                           name='median'))
fig2.show()

**4. Описать данные**

1.1 Данный ресурс предоставляет 15 курсов различной направленности в таких сферах, как программирование, дизайн, диджитал-маркетинг.

1.2 Каждый курс разделен на модули. Каждый модуль состоит из уроков. Больше всего модулей на курсах: "UX-дизайн"(20шт), "Анимация интерфейса"(21), "Веб-дизайн с нуля"(19) и "Веб-разработчик"(20). Самым наименьшим количеством обладают курсы: "Excel Базовый"(9шт), "Java-разработчик с нуля"(9), "PHP-разработчик с 0 до PRO"(8), "Веб-верстка для начинающих"(8).

1.3 По количеству уроков людируют модули  "Email-маркетинг" на курсе "Интернет-маркетолог от Ingate" (29шт); "Числа, строки и даты" на курсе "Java-разработчик с нуля" (27 шт).
Количество модулей всего с одним уроком является наибольшим среди всех (34шт).

1.4 Медианное количество уроков в модуле на каждом курсе варьируется от 2 до 10.  На курсе "Java-разработчик с нуля" в модулях в среднем больше уроков, чем на остальных курсах. Курс "Веб-разработчик" имеет наименьший этот показатель, равный двум. 
В лидирах курсы, связанные с языком программирования Java. С меньшим значением медианы курсы дизайна, среди которых половина для новичков.

1.5 Самый популярные курсы: "Интернет-маркетолог" - 2168 студентов, "Веб-дизайн с нуля - 2014, "Веб-верстка для начинающих" - 2004. Самые непопулярные курсы: "SMM-маркетолог" - 511 студентов, "Веб-дизайн с нуля" - 518. Заметно, что почти все курсы в направлении дизайна в лидерах. Программирование в середичках.

1.6 Некоторые студенты при регистрации либо записывают неправильные или ложные данные, такие как дата рождения и т.п., либо вообще их неуказывают.
По отфильтрованным данным даты рождения можно сделать вывод, что на курсы регистрируются люди, как минимального возраста, так и максимального, если, конечно данные верные. Средняя и медиана показывают, что 29 возраст преобладает. Аудитория более зрелая, скорее всего люди стремятся получить дополнительное образование.

1.7 Если рассматривать предыдущие показатели в разрезе курсов, то максимальная медиана на курсах "Excel Базовый" 34 года, "Руководитель digital-проектов" 32 года и "Интернет-маркетолог от Ingate" 31 год. 
Больше всего 28 летних. По метрике минимум и максимум возраста выводу будут не столь корректны из-за недостоверности указываемых данных.

2.Эту информацию мы уже рассмотрели выше. График лишь более наглядно демонстрирует распределение данных.

3.На этом графике уже более наглядно вырисовываются тенденции. Так целых 7 курсов 
имеют показатели количества студентов выше общей медианы. По цветам сразу видно, что 3 курса относятся к Development (красный цвет) и к Design (цвет фиолетовый).

## 2. Расчет потенциальной нагрузки на преподавателей

1. **(p)** Рассчитать прирост студентов на каждом курсе в каждом месяце за всю историю (каждый месяц в диапазоне от марта 2016 до июля 2019 включительно). Считать дату начала прохождения курса студентом по дате начала первой домашней работы.
2. **(m)** На основании первого пункта построить line-graph с приростом студентов в каждом месяце для каждого курса. 15 графиков. Графики должны иметь заголовки, оси должны быть подписаны. Ticks нужно развернуть так, чтобы они были читаемы.
3. **(m)** На основании первого пункта построить line-graph с несколькими линиями, отражающими прирост студентов в каждом месяце для каждого курса. 15 линий на графике. Ticks нужно развернуть так, чтобы они были читаемы. График должен иметь заголовок. Ось, отражающая прирост, должна быть подписана. Линия для каждого курса должна иметь свой цвет (нужна легенда). Рамок у графика быть не должно ⭐
4. **(p)** Рассчитать количество прогрессов по выполнению домашних работ в каждом месяце за всю историю (каждый месяц в диапазоне от марта 2016 до июля 2019 включительно) для каждого курса. Учитывать, что выполнение домашнего задания может перетекать из одного месяца в другой (такие дз надо включать в общее число прогрессов для всех месяцев, которые покрывает срок выполнения этих дз)
5. **(m)** Построить line-graph по четвертому пункту. 15 графиков. Графики должны иметь заголовки, оси должны быть подписаны. Ticks нужно развернуть так, чтобы они были читаемы
6. **(m)** Построить один line-graph для всех курсов по четвертому пункту. 15 линий на графике. Ticks нужно развернуть так, чтобы они были читаемы. График должен иметь заголовок. Ось, отражающая количество прогрессов, должна быть подписана. Линия для каждого курса должна иметь свой цвет (нужна легенда). Рамок у графика быть не должно ⭐
7. На основании рассчитанных значений сделайте аналитический вывод (должен быть полным и покрывать все полученные выше метрики)

**Задание №1**

In [5]:
import datetime

# принудительно определяем тип столбца "start_date", как datetime.
df['start_date'] = pd.to_datetime(df['start_date'], errors='coerce')

In [142]:
# фильтруем датафрейм по условиям задания
df_new_student = df[(df['status'] == 'start') & (df['is_homework'] == True) &
                    (df['start_date'] >= '2016-03-01') & (df['start_date'] < '2019-08-01')]

In [175]:
# группируем данные по названию курса, id студента и определяем дату начала выполнения первой домашней работы
df_new_student_first = df_new_student.groupby(['course_title', 'student_id'], as_index=False).start_date.agg({'date': 'min'})
# считаем количество студентов для каждого курса и месяца
df_new_student_first_count = df_new_student_first.groupby(['course_title',
                                                           pd.Grouper(key='date', freq='M')]).agg({'student_id': 'count'}).reset_index()
df_new_student_first_count.set_index('date', inplace=True)
# повышаем дискретизацию, с частотой месяц, так как есть месяцы в которых не было новых студентов (количество
# заполняем 0
df_new_student_first_resample = df_new_student_first_count.groupby('course_title').resample('M').first().fillna(0)

**Задание №2**

In [180]:
import plotly.express as px

dfn = df_new_student_first_resample
fig1 = px.area(dfn, x=dfn.index.get_level_values(1), y=dfn.student_id,
               facet_col=dfn.index.get_level_values(0), facet_col_wrap=2,
               labels=dict(x='Дата', student_id='Прирост', facet_col='Курс', color='Курс'),
               color=dfn.index.get_level_values(0),
               template='plotly_white')
fig1.update_layout(height=1800, width=800, showlegend=False, margin=dict(l=0, r=0),
                   title=dict(text='Прирост студентов в каждом месяце для каждого курса', x=0.5))
fig1.update_xaxes(fixedrange=True, tickangle=45, title_font={'size': 15}, title_standoff=15, showticklabels=True,
                  nticks=12, tickfont_size=10)
fig1.update_yaxes(fixedrange=True, showticklabels=True, tickfont_size=8)
fig1.show()

**Задание №3**

In [181]:
fig2 = px.line(dfn, x=dfn.index.get_level_values(1), y=dfn.student_id,
               labels=dict(x='Дата', student_id='Прирост', course_title='Курс', color='Курс'),
               color=dfn.course_title,
               title='Прирост студентов в каждом месяце для каждого курса',
               template='plotly_white')
fig2.update_layout(legend_orientation='h',
                   title=dict(text='Прирост студентов в каждом месяце для каждого курса', x=0.5),
                   height=600, legend_y=-0.3,
                   )
fig2.update_xaxes(tickangle=45, title_font={'size': 15}, title_standoff=15,
                  nticks=24, tickfont_size=9,)
fig2.update_yaxes(tickfont_size=9)
fig2.show()

**Задание №4**

4. **(p)** Рассчитать количество прогрессов по выполнению домашних работ в каждом месяце за всю историю (каждый месяц
в диапазоне от марта 2016 до июля 2019 включительно) для каждого курса. Учитывать, что выполнение домашнего задания
может перетекать из одного месяца в другой (такие дз надо включать в общее число прогрессов для всех месяцев, которые
покрывает срок выполнения этих дз)

In [280]:
# новая таблица с фильтром по условию задания
df_homework = df[(df.status == 'done') & (df.is_homework == True) & (df['start_date'] >= '2016-03-01') &
                 (df['start_date'] < '2019-08-01')][['course_title', 'start_date', 'finish_date']]

In [305]:
import numpy as np

df_homework.start_date = pd.to_datetime(df_homework.start_date)
df_homework.finish_date = pd.to_datetime(df_homework.finish_date)

n = pd.DatetimeIndex(df_homework.start_date)
m = pd.DatetimeIndex(df_homework.finish_date)

c = n.month != m.month
# month = m.month - n.month
# df_homework['difference'] = month
df_homework = df_homework[c]
df_homework['diff'] = df_homework.finish_date.to_numpy(dtype="datetime64[M]") - df_homework.start_date.to_numpy(dtype="datetime64[M]")
# new.start_date.iloc[0] + np.timedelta64(1,'M')
# np.max(new.difference)
# (np.datetime64('2019-04') - np.datetime64('2019-04')) == 0
# new.difference.values

<ipython-input-305-31935dca986b>:13: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



c:\users\ilya\pyenv\coursework1\lib\site-packages\pandas\core\generic.py:5489: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



course_title start_date finish_date    diff
3565         Веб-дизайн PRO 2.0 2019-04-01  2019-05-01 30 days
3566         Веб-дизайн PRO 2.0 2019-02-01  2019-04-01 59 days
3571         Веб-дизайн PRO 2.0 2019-04-01  2019-06-01 61 days
3575         Веб-дизайн PRO 2.0 2019-05-01  2019-06-01 31 days
3576         Веб-дизайн PRO 2.0 2019-02-01  2019-03-01 28 days
...                         ...        ...         ...     ...
350643  Java-разработчик c нуля 2019-06-01  2019-07-01 30 days
350649  Java-разработчик c нуля 2019-06-01  2019-07-01 30 days
350652  Java-разработчик c нуля 2019-06-01  2019-07-01 30 days
350655  Java-разработчик c нуля 2019-06-01  2019-07-01 30 days
350656  Java-разработчик c нуля 2019-06-01  2019-07-01 30 days

[14556 rows x 4 columns]

## 3. Выявление проблемных модулей

1. **(p)** Рассчитать минимальное, максимальное, среднее, медианное время прохождения каждого модуля (разность между временем начала и окончания выполнения домашней работы) для каждого курса. Если домашних заданий в модуле несколько, то считать разность между временем начала выполнения первой домашней работы и временем окончания выполнения последней домашней работы в модуле
2. **(m)** На основании первого пункта построить line-graph с медианным временем прохождения каждого модуля для каждого курса. 15 графиков. Графики должны иметь заголовки
3. **(p)** Чтобы выявить сезонность, посчитать медианное время выполнения домашней работы по месяцам (12 месяцев, январь-декабрь) для каждого курса.
4. **(m)** На основании третьего пункта построить line-graph, на который будут нанесены линии для каждого курса с медианным временем выполнения домашней работы по месяцам. 15 линий на графике. График должен иметь заголовок. Ось, отражающая время прохождения, должна быть подписана. Линия для каждого курса должна иметь свой цвет (нужна легенда). Рамок у графика быть не должно  ⭐
5. На основании рассчитанных значений сделайте аналитический вывод (должен быть полным и покрывать все полученные выше метрики)

**1.**




## 4. Расчет конверсии

1. **(p)** Посчитать конверсию перехода студентов из одного модуля в другой на каждом курсе. Формула: отношение количества студентов, приступивших к выполнению домашнего задания в этом модуле (если дз в модуле несколько, то считать по первому дз в модуле), к количеству студентов, сдавших задание в предыдущем модуле (если дз в модуле несколько, то считать по последнему дз в модуле).
2. **(m)** Постройте bar-chart, отражающий конверсию перехода студентов из одного модуля в другой на каждом курсе. График должен иметь заголовок. Ticks нужно развернуть так, чтобы они были читаемы
3. **(m)** Постройте горизонтальный (столбцы должны располагаться горизонтально) bar-chart, отражающий конверсию перехода студентов из одного модуля в другой на каждом курсе. 15 графиков. Графики должны иметь заголовки. Ticks должны содержать номер и название модуля. Цвет столбцов графиков должен содержать информацию о сфере, к которой относится курс (нужна легенда). Прозрачность должна стоять на отметке 0.1. На графики должна быть нанесена линия медианы конверсии для каждого курса. У медианы должен быть свой цвет. Рамок у графиков быть не должно ⭐
4. На основании рассчитанных значений сделайте аналитический вывод (должен быть полным и покрывать все полученные выше метрики)

In [ ]:
### YOUR CODE HERE ###




## 5. Метрика успеваемости 

&nbsp;&nbsp;&nbsp;&nbsp;*(необязательное задание)*

Иногда студенты берут курсы, которые оказываются для них неподъемными. Это может быть как по причинам недостаточной изначальной подготовки, так и по причинам, связанным с низкой мотивацией студента. Для улучшения качества контента полезно выявить причину. На основании имеющихся данных придумайте метрику успеваемости студента. **Обоснуйте ее.** Выявите таких студентов на каждом курсе, чтобы предоставить продюсерам список проблемных студентов. 

In [ ]:
### YOUR CODE HERE ###


